core/finality-grandpa/src/finality_proof.rs

```diff
@@ -0,0 +1,428 @@
+//! 3) proof-of-execution of the `grandpa_authorities` call at the block F.
+//!
+//! Since earliest possible justification is returned, the GRANDPA authorities set
+//! at the block F is guaranteed to be the same as in the block B (this is because block
+//! that enacts new GRANDPA authorities set always comes with justification). It also
+//! means that the `set_id` is the same at blocks B and F.
+//!
+//! The caller should track the `set_id`. The most straightforward way is to fetch finality
+//! proofs ONLY for blocks on the tip of the chain and track the latest known `set_id`.
+
+use std::collections::HashMap;
+
+use client::{
+	blockchain::Backend as BlockchainBackend,
+	error::{Error as ClientError, ErrorKind as ClientErrorKind, Result as ClientResult},
+	light::fetcher::RemoteCallRequest,
+};
+use codec::{Encode, Decode};
+use grandpa::BlockNumberOps;
+use runtime_primitives::generic::BlockId;
+use runtime_primitives::traits::{
+	NumberFor, Block as BlockT, Header as HeaderT, One,
+};
+use substrate_primitives::{Ed25519AuthorityId, H256};
+
+use GrandpaJustification;
+
+/// Prepare proof-of-finality for the given block.
+///
+/// The proof is the serialized `FinalityProof` constructed using earliest known
+/// justification of the block. None is returned if there's no known justification atm.
+pub fn prove_finality<Block: BlockT, B, G>(
+	blockchain: &B,
+	generate_execution_proof: G,
+	block: Block::Hash,
+) -> ::client::error::Result<Option<Vec<u8>>>
+	where
+		B: BlockchainBackend<Block>,
+		G: Fn(&BlockId<Block>, &str, &[u8]) -> ClientResult<Vec<Vec<u8>>>,
+{
+	let block_id = BlockId::Hash(block);
+	let mut block_number = blockchain.expect_block_number_from_id(&block_id)?;
+
+	// early-return if we sure that the block isn't finalized yet
+	let info = blockchain.info()?;
+	if info.finalized_number < block_number {
+		return Ok(None);
+	}
+
+	// early-return if we sure that the block is NOT a part of canonical chain
+	let canonical_block = blockchain.expect_block_hash_from_id(&BlockId::Number(block_number))?;
+	if block != canonical_block {
+		return Err(ClientErrorKind::Backend(
+			"Cannot generate finality proof for non-canonical block".into()
+		).into());
+	}
+
+	// now that we know that the block is finalized, we can generate finalization proof
+
+	// we need to prove grandpa authorities set that has generated justification
+	// BUT since `GrandpaApi::grandpa_authorities` call returns the set that becames actual
+	// at the next block, the proof-of execution is generated using parent block' state
+	// (this will fail if we're trying to prove genesis finality, but such the call itself is redundant)
+	let mut current_header = blockchain.expect_header(BlockId::Hash(block))?;
+	let parent_block_id = BlockId::Hash(*current_header.parent_hash());
+	let authorities_proof = generate_execution_proof(
+		&parent_block_id,
+		"GrandpaApi_grandpa_authorities",
```

Comment:
unrelated @bkchr @cheme is it possible to have the `decl_runtime_apis!` export some constants?